course4/lecture4_en.md

```diff
@@ -0,0 +1,306 @@
+
+* type equality implying isomorphism
+* type multiplication forming product types (tuples or structures)
+* type addition forming sum types (enumerated types) 
+
+Here, **Zero** is a type that corresponds to an **empty type**. We can define an empty enumerated type without any cases; such a type has no constructors, and no values can be constructed, making it empty. **One** corresponds to a type with only one element, which we call the **Unit type**, and its value is a zero-tuple.
+
+Let's verify the properties mentioned earlier: any number multiplied by 1 equals itself, and any number plus 0 equals itself.
+
+```moonbit
+fn f[T](t: T) -> (T, Unit) { (t, ()) }
+fn g[T](pair: (T, Unit)) -> T { pair.0 }
+```
+
+In this context, a type `T` multiplied by 1 implies that `(T, Unit)` is isomorphic to `T`. We can establish a set of mappings: it's straightforward to go from `T` to `(T, Unit)` by simply adding the zero-tuple. Conversely, going from `(T, Unit)` to `T` involves ignoring the zero-tuple. You can intuitively find that they are isomorphic. 
+
+```moonbit
+enum PlusZero[T] { CaseT(T); CaseZero(Nothing) }
+
+fn f[T](t: PlusZero) -> T {
+  match t {
+    CaseT(t) => t
+    CaseZero(_) => abort("Impossible case, no such value.")
+  }
+}
+
+fn g[T](t: T) -> PlusZero { CaseT(t) }
+```
+
+The property of any type plus zero equals itself means that, for any type, we define an enumerated type `PlusZero`. One case contains a value of type `T`, and the other case contains a value of type `Nothing`. This type is isomorphic to `T`, and we can construct a set of mappings. Starting with `PlusZero`, we use pattern matching to discuss the cases. If the included value is of type `T`, we map it directly to `T`. If the type is `Nothing`, this case will never happen because there are no values of type `Nothing`, so we use `abort` to handle, indicating that the program will terminate. Conversely, we only need to wrap `T` with `CaseT`. It's essential to emphasize that this introduction is quite basic, providing an intuitive feel. Explore further if you are interested.
+
+Here are a few examples. 
+
+```moonbit
+enum Coins { Head; Tail }
+```
+
+$\texttt{Coins} = 1 + 1 = 2$
+
+```moonbit
+enum DaysOfWeek { Monday; Tuesday; ...; }
+```
+
+$\texttt{DaysOfWeek} = 1 + 1 + 1 + 1 + 1 + 1 + 1 = 7$
+
+The data type for the coin toss can be considered as 1 + 1, as each case, `Head` and `Tail`, actually represents a set with only one value. Therefore, each case is isomorphic to the Unit type. When combined by the sum type, the `Coin` type becomes 1 + 1 = 2, representing a set with two values, which is isomorphic to any other type with two values. Similarly, `DaysOfWeek` represents a set of seven values, isomorphic to any other type with seven values. 
+
+A more interesting example is `List`, using `List[Int]` as an example. 
+
+$$
+\begin{array}{rl}
+\texttt{enum} \;\; \texttt{List} & = \texttt{Nil} + \texttt{Int} \times \texttt{List} \\  
```

Comment:
I think the preview of github does not support `\;`. Is there any better alternative?

Reply:
Just `\ ` is OK. 

$$
\begin{array}{rl}
\texttt{enum} \ \ \texttt{List} & = \texttt{Nil} + \texttt{Int} \times \texttt{List} \\  
& = \texttt{1} + \texttt{Int} \times \texttt{List} \\
& = \texttt{1} + \texttt{Int} \times (\texttt{1} + \texttt{Int} \times \texttt{List} ) \\
& = \texttt{1} + \texttt{Int} \times \texttt{1} + \texttt{Int} \times \texttt{Int} \times \texttt{List} \\
& = \texttt{1} + \texttt{Int} + \texttt{Int} \times \texttt{Int} + \texttt{Int} * \texttt{Int} \times \texttt{Int} + ...
\end{array}
$$